p2p/p2p.go

```diff
@@ -99,86 +114,193 @@ func (s *Service) Stop() error {
+		s.blockAnnRec[msg.Id()] = true
 	case TransactionMsgType:
 		if s.txMessageRec[msg.Id()] {
-			return nil
+			return false
 		}
 		s.txMessageRec[msg.Id()] = true
 	default:
-		log.Error("Invalid message type", "type", msgType)
-		return nil
-	}
-
-	encodedMsg, err := msg.Encode()
-	if err != nil {
-		return err
+		// status message type not valid
+		return false
 	}
 
-	s.host.broadcast(encodedMsg)
-
-	return err
+	return true
 }
 
-// handleStream handles the stream, and rebroadcasts the message based on it's type
+// handleStream parses the message written to the data stream and calls the
+// associated message handler (status or non-status) based on message type
 func (s *Service) handleStream(stream net.Stream) {
+
+	// parse message and return on error
 	msg, _, err := parseMessage(stream)
 	if err != nil {
+		log.Debug("parse message", "error", err)
 		return
 	}
 
-	// Write the message back to channel
-	s.msgSend <- msg
+	log.Debug(
+		"handle stream",
+		"host", stream.Conn().LocalPeer(),
+		"peer", stream.Conn().RemotePeer(),
+		"protocol", stream.Protocol(),
+		"message", msg,
+	)
+
+	if msg.GetType() == StatusMsgType {
+		// handle status message type
+		s.handleStreamStatus(stream, msg)
+	} else {
+		// handle other message types
+		s.handleStreamNonStatus(stream, msg)
+	}
+
+}
+
+// handleStreamStatus handles status messages written to the stream
+func (s *Service) handleStreamStatus(stream network.Stream, msg Message) {
+
+	// TODO: use generated status message
+	hostStatus := statusMessage
+
+	switch {
+
+	case hostStatus.String() == msg.String():
+		log.Debug(
+			"status match",
+			"host", stream.Conn().LocalPeer(),
+			"peer", stream.Conn().RemotePeer(),
+		)
+
+		// TODO: store status in peer metadata
+		s.host.peerStatus[stream.Conn().RemotePeer()] = true
+
+	default:
+		log.Debug(
+			"status mismatch",
+			"host", stream.Conn().LocalPeer(),
+			"peer", stream.Conn().RemotePeer(),
+		)
+
+		// TODO: store status in peer metadata
+		s.host.peerStatus[stream.Conn().RemotePeer()] = false
+
+		// TODO: drop peer if status mismatch
 
-	// Rebroadcast all messages except for status messages
-	if msg.GetType() != StatusMsgType {
-		err = s.Broadcast(msg)
-		if err != nil {
-			log.Debug("failed to broadcast message: ", err)
-			return
-		}
 	}
 }
 
-var _ module.P2pApi = &Service{}
+// handleStreamNonStatus handles non-status messages written to the stream
+func (s *Service) handleStreamNonStatus(stream network.Stream, msg Message) {
+
+	// TODO: get peer status from peer metadata
+	status := s.host.peerStatus[stream.Conn().RemotePeer()]
+
+	// return if status message has not been confirmed
+	if !status {
+		log.Debug(
+			"message ignored",
+			"host", stream.Conn().LocalPeer(),
+			"peer", stream.Conn().RemotePeer(),
+			"protocol", stream.Protocol(),
+			"message", msg,
+		)
+		return
+	}
 
-// ID returns the host's ID
+	// check and store message, returns true if valid new message
+	if !s.verifyNewMessage(msg) {
```

Comment:
naming may be a bit confusing, the message could still be valid but seen already. maybe change to `shouldBroadcast` or something?

Reply:
Good point. I can update.